test/Templates.UnitTests/EncodeTests.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using Xunit;
+using System;
+
+namespace Templates.UnitTests
+{
+    public class EncodeTests
+    {
+       [Fact]
+        public void AllFilesInTemplatesShouldBeUTF8Encoded()
+        {
+            var dirPath = Path.Combine(AppContext.BaseDirectory, "TestAssets", "Templates");
+            CheckAllFilesUTF8Encoded(dirPath);
+        }
+
+        private void CheckAllFilesUTF8Encoded(string dirPath)
+        {
+            foreach (var filePath in Directory.GetFiles(dirPath))
+            {
+                if (Directory.Exists(filePath))
+                {
+                    CheckAllFilesUTF8Encoded(filePath);
+                }
+                else if (File.Exists(filePath))
+                {
+                    Assert.True(IsUTF8EncodedFile(filePath), $"{filePath} should be UTF-8 encoded");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Check first three bytes to check the file is UFT-8 or not.
+        /// First three bytes should be EF BB BF for UTF-8
+        /// Ref: https://msdn.microsoft.com/en-us/library/windows/desktop/dd374101(v=vs.85).aspx
+        /// </summary>
+        /// <param name="filePath"></param>
+        /// <returns>true; If file encoding is UTF-8</returns>
+        private bool IsUTF8EncodedFile(string filePath)
+        {
+            var bom = new byte[3];
+            using (var file = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+            {
+                file.Read(bom, 0, 3);
+            }
+
+            return bom[0] == 0xEF && bom[1] == 0xBB && bom[2] == 0xBF;
```

Comment:
The name of the functions and the error messages aren't accurate. This only checks if the file has the *optional* UTF-8 BOM, but does not validate that its byte sequence is valid UTF-8 (perhaps we should also check that, see below). 

You could check that the file is valid UTF-8 using:
``` C#
var strictUTF8 = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true, throwOnInvalidBytes: true);
using (var reader = new StreamReader(filePath, strictUTF8, detectEncodingFromByteOrderMarks: false))
{
    try
    {
        reader.ReadToEnd();
    }
    catch (DecoderFallbackException)
    {
        return false;
    }
}
```

and then also check that the file has the BOM as you are now.

Then I'd rename to IsUTF8EncodedWithBOM and adjust error messages accordingly too.